### PR TITLE
Allow schedule on ZStream.repeatEffectWith rely on effect value

### PR DESCRIPTION
**Pull request by luis3m**
Resolves https://github.com/zio/zio/issues/3598

### CODE REVIEW

**Comment by simpadjo** [comment]
Hey! Semantics of  `repeatWith` has changed. It used to update the schedule at the end of every chunk and now it updates it only if the stream was nonempty. No surprise, since any stream can produce `Any` but not every stream can produce `A`.

I think that `repeatEffectWith` is a better place to make changes. Unlike `repeatWith`, here we know that a substream emitted exactly one value and can pass the value to the schedule w/o losing any information about the stream.

**Comment by iravid** [comment]
Maybe repeatWith’s schedule can take a Chunk[O] as input?

**Comment by simpadjo** [comment]
I thought about it as well, but FWIW it exposes chunking to high level API.

**Comment by luis3m** [comment]
@simpadjo @iravid done

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -3479,11 +3479,10 @@ object ZStream extends ZStreamPlatformSpecificConstructors {
...
    * Creates a stream from an effect producing a value of type `A` which repeats using the specified schedule
    */
-  def repeatEffectWith[R, E, A](
-    fa: ZIO[R, E, A],
-    schedule: Schedule[R, Any, _]
-  ): ZStream[R, E, A] =
-    fromEffect(fa).repeat(schedule)
+  def repeatEffectWith[R, E, A](effect: ZIO[R, E, A], schedule: Schedule[R, A, _]): ZStream[R, E, A] =
+    fromEffect(schedule.initial).flatMap { initial =>
+      unfoldM(initial)(state => effect.flatMap(value => schedule.update(value, state).option.map(_.map((value, _)))))
```
Comment (simpadjo) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:3484]: Replacing `.option` with `.fold` will kill 2 `.map`s

File: streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala [hunk 2]
```diff
@@ -2804,18 +2804,30 @@ object ZStreamSpec extends ZIOBaseSpec {
...
+                        .take(2)
+                        .runDrain
+                        .fork
+              _      <- TestClock.adjust(50.millis)
+              _      <- fiber.join
+              result <- ref.get
+            } yield assert(result)(equalTo(List(1, 1)))
+          ),
+          testM("allow schedule rely on effect value")(checkM(Gen.int(1, 100)) { (length: Int) =>
+            var x        = length
```
Comment (simpadjo) [inline streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala:2822]: If you want to countdown from `length` to `0` you can, for example, allocate a `Ref[Int]` and use `ref.getAndUpdate`.

File: streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala [hunk 3]
```diff
@@ -2804,18 +2804,30 @@ object ZStreamSpec extends ZIOBaseSpec {
...
+            val effect   = UIO(if (x <= 0) 0 else { val y = x; x -= 1; y })
+            val schedule = Schedule.identity.addDelay((i: Int) => (if (i == 0) 1 else 0).seconds).forever
+            for {
```
Comment (simpadjo) [inline streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala:2825]: As far as I can see, this test doesn't check timings despite it uses Clock.
I feel that we can achieve the same coverage with a simpler testcase.
What about, e.g.:
1) take an effect that counts down from 100 to 0, but crashes on 51
2) pick a schedule that stops when it sees `50`
3) check that it collected (100..50)
4) optionally the schedule might as well crash on elements below 49

**Comment by simpadjo** [comment]
Looks good! But let's try to simplify the test.

File: streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala (hunk 2 above)
Comment (luis3m) [reply streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala]: Right, should've thought about that

File: streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala (hunk 3 above)
Comment (luis3m) [reply streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala:2822]: @simpadjo let me give that a try

Comment (luis3m) [reply streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala:2822]: @simpadjo thanks for the suggestions, done.

**Comment by luis3m** [comment]
@simpadjo test has been simplified as suggested

**Comment by simpadjo** [comment]
@luis3m great work!

The last comment: if PR description was "Closes #3598" instead of "Addresses #3598", merging this PR would authomatiacally close the issue. It's good to keep the issue list clean.

**Comment by luis3m** [comment]
@simpadjo oh, I didn't know that. Good to know and thanks :)

**Pull request merged**